src/main/java/bio/terra/service/dataset/StorageResource.java

```diff
@@ -1,92 +1,40 @@
 package bio.terra.service.dataset;
 
-import bio.terra.app.model.GoogleCloudResource;
-import bio.terra.app.model.GoogleRegion;
+import bio.terra.app.model.CloudRegion;
+import bio.terra.app.model.CloudResource;
 import bio.terra.model.CloudPlatform;
 import bio.terra.model.StorageResourceModel;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
 
-import java.util.Objects;
 import java.util.UUID;
 
-public class StorageResource {
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "cloudPlatform")
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = AzureStorageResource.class, name = "azure"),
+    @JsonSubTypes.Type(value = GoogleStorageResource.class, name = "gcp")
+})
+public interface StorageResource<Region extends CloudRegion, Resource extends CloudResource> {
```

Comment:
I'm not sure this is a great idea, to make this a generic type.

That being said, if you do this, you should make this abstract and move all the fields back here, except for `CloudPlatform` which would still be `abstract`. 

You can use the type parameters to the class to declare the fields, the constructor args, and the getter/setter types.

Reply:
Oh, and if you do that, then you can definitely move equals/tostring/hashcode back to `StorageResource`

Reply:
Huh..interesting..let me play with that

Reply:
And fwiw, without generics, it wasn't even test failing but straight up not compiling

Reply:
Ugh...I tried the abstract route but...enums cant extend classes so no luck.  So we're either doing generics or putting casts everywhere...(unless there's another option I'm missing)

Reply:
Wait scratch that last comment...wrong class